File: codes/scala/chapter_backtracking/preorder_traversal_ii_compact.scala
import scala.collection.mutable.ListBuffer

/**
 * File: preorder_traversal_ii_compact.scala
 * Author: WangBlue (wangjiping596@gmail.com)
 * Create: 2024-08-16 11:10
 */

class preorder_traversal_ii_compact {
  var path = new ListBuffer[TreeNode]()
  var res = new ListBuffer[ListBuffer[TreeNode]]()

  /* 前序遍历：例题二 */
  def preOrder(root: TreeNode): Unit = {
    if (root == null) {
      return
    }
    // 尝试
    path = path :+ root
    if (root.value == 7) {
      // 记录解
      res = res :+ path
    }
    preOrder(root.left)
    preOrder(root.right)
    // 回退
    path = path.dropRight(1)
  }
}

object preorder_traversal_ii_compact {

  def main(args: Array[String]): Unit = {
    val list: List[Option[Int]] = List(1, 7, 3, 4, 5, 6, 7).map(Option(_))
    val root: TreeNode = new TreeNode().listToTree(list)
    println("\n初始化二叉树")
    new PrintUtil().printTree(root)

    // 前序遍历
    val preor = new preorder_traversal_ii_compact()
    preor.preOrder(root)
    println("\n输出所有根节点到节点 7 的路径")
    preor.res.foreach { path =>
      val vals = path.map(_.value)
      println(vals)
    }
  }
}

File: ru/codes/csharp/chapter_backtracking/preorder_traversal_ii_compact.cs
/**
 * File: preorder_traversal_ii_compact.cs
 * Created Time: 2023-04-17
 * Author: hpstory (hpstory1024@163.com)
 */

namespace hello_algo.chapter_backtracking;

public class preorder_traversal_ii_compact {
    List<TreeNode> path = [];
    List<List<TreeNode>> res = [];

    /* Предварительный обход: пример 2 */
    void PreOrder(TreeNode? root) {
        if (root == null) {
            return;
        }
        // Попытка
        path.Add(root);
        if (root.val == 7) {
            // Записать решение
            res.Add(new List<TreeNode>(path));
        }
        PreOrder(root.left);
        PreOrder(root.right);
        // Откат
        path.RemoveAt(path.Count - 1);
    }

    [Test]
    public void Test() {
        TreeNode? root = TreeNode.ListToTree([1, 7, 3, 4, 5, 6, 7]);
        Console.WriteLine("\nИнициализация двоичного дерева");
        PrintUtil.PrintTree(root);

        // Предварительный обход
        PreOrder(root);

        Console.WriteLine("\nВсе пути от корня к узлу 7");
        foreach (List<TreeNode> path in res) {
            PrintUtil.PrintList(path.Select(p => p.val).ToList());
        }
    }
}
